src/main.rs

```diff
@@ -232,6 +233,89 @@ fn main() {
+///                   using the API socket.
+pub fn start_vmm(
+    api_shared_info: Arc<RwLock<InstanceInfo>>,
+    api_event_fd: EventFd,
+    from_api: Receiver<Box<(VmmAction, OutcomeSender)>>,
+    seccomp_level: u32,
+    config_json: Option<String>,
+) {
+    // If this fails, consider it fatal. Use expect().
+    let mut vmm =
+        Vmm::new(api_shared_info, &api_event_fd, seccomp_level).expect("Cannot create VMM");
+
+    if let Some(json) = config_json {
+        vmm.configure_from_json(json).unwrap_or_else(|err| {
+            error!(
+                "Setting configuration for VMM from one single json failed: {}",
+                err
+            );
+            process::exit(i32::from(vmm::FC_EXIT_CODE_BAD_CONFIGURATION));
+        });
+        vmm.start_microvm().unwrap_or_else(|err| {
+            error!(
+                "Starting microvm that was configured from one single json failed: {}",
+                err
+            );
+            process::exit(i32::from(vmm::FC_EXIT_CODE_UNEXPECTED_ERROR));
+        });
+        info!("Successfully started microvm that was configured from one single json");
+    }
+
+    let exit_code = loop {
+        match vmm.run_event_loop() {
+            Err(e) => {
+                error!("Abruptly exited VMM control loop: {:?}", e);
+                break vmm::FC_EXIT_CODE_GENERIC_ERROR;
+            }
+            Ok(exit_reason) => match exit_reason {
+                EventLoopExitReason::Break => {
+                    info!("Gracefully terminated VMM control loop");
+                    break vmm::FC_EXIT_CODE_OK;
+                }
+                EventLoopExitReason::ControlAction => {
+                    if let Err(e) = api_event_fd.read() {
+                        error!("Error reading VMM API event_fd {:?}", e);
+                        break vmm::FC_EXIT_CODE_GENERIC_ERROR;
+                    } else {
+                        match from_api.try_recv() {
+                            Ok(t) => {
+                                let t = *t;
+                                let (action_request, sender) = (t.0, t.1);
+                                // Run the requested action and send back the result.
+                                sender
+                                    .send(vmm.run_vmm_action(action_request))
+                                    .map_err(|_| ())
+                                    .expect("one-shot channel closed");
+                            }
+                            Err(TryRecvError::Empty) => {
+                                warn!("Got a spurious notification from api thread");
+                            }
+                            Err(TryRecvError::Disconnected) => {
+                                panic!("The channel's sending half was disconnected. Cannot receive data.");
+                            }
+                        };
+                    }
```

Comment:
I would move this code to a separate function that handles API requests to make the loop more clear and readable.

Reply:
Sounds good. Will do after I get more feedback so I can make multiple changes at once.